connectors/sources/mssql.py

```diff
@@ -0,0 +1,126 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+"""Microsoft SQL source module is responsible to fetch documents from Microsoft SQL."""
+from sqlalchemy import create_engine
+from sqlalchemy.engine import URL
+
+from connectors.sources.generic_database import GenericBaseDataSource
+
+SECURED_CONNECTION = False
+# Below schemas are system schemas and the tables of the systems schema's will not get indexed
+SYSTEM_SCHEMA = [
+    "INFORMATION_SCHEMA",
+    "db_owner",
+    "db_accessadmin",
+    "db_securityadmin",
+    "db_ddladmin",
+    "db_backupoperator",
+    "db_datareader",
+    "db_datawriter",
+    "db_denydatareader",
+    "db_denydatawriter",
+    "sys",
+]
+
+QUERIES = {
+    "PING": "SELECT 1+1",
+    "ALL_TABLE": "SELECT table_name FROM information_schema.tables WHERE TABLE_SCHEMA = '{schema}'",
+    "TABLE_PRIMARY_KEY": "SELECT C.COLUMN_NAME FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS T JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE C ON C.CONSTRAINT_NAME=T.CONSTRAINT_NAME WHERE C.TABLE_NAME='{table}' and C.TABLE_SCHEMA='{schema}' and T.CONSTRAINT_TYPE='PRIMARY KEY' ",
+    "TABLE_DATA": 'SELECT * FROM {schema}."{table}"',
+    "TABLE_LAST_UPDATE_TIME": "SELECT last_user_update FROM sys.dm_db_index_usage_stats WHERE object_id=object_id('{schema}.{table}')",
+    "TABLE_DATA_COUNT": 'SELECT COUNT(*) FROM {schema}."{table}"',
+    "ALL_SCHEMAS": "SELECT s.name from sys.schemas s inner join sys.sysusers u on u.uid = s.principal_id",
+}
+
+
+class MSSQLDataSource(GenericBaseDataSource):
+    """Microsoft SQL Server"""
+
+    def __init__(self, configuration):
+        """Setup connection to the Microsoft SQL database-server configured by user
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+        self.mssql_driver = self.configuration["mssql_driver"]
+        self.secured_connection = self.configuration["secured_connection"]
+        self.connection_string = ""
+        self.queries = QUERIES
```

Comment:
Why do we store a constant in `self.queries`? Do we reassign `self.queries` anywhere or allow it via the API of this data source?

Reply:
`QUERIES` constant is different for each dialect, so when we run the connector data source file will be called first. that's why we have initialized the `self.queries` in each data source file and later used it in the generic_database file.

Reply:
If `QUERIES` always has the same structure/keys you could create a `Queries` class providing a clear interface, what types of queries `generic_database.py` expects. By using an abstract class and its capabilities enforcing a specific API it's more obvious what queries you need to provide in an concrete implementation:

`
class Queries(ABC):

  @abstractmethod
  def ping(self):
    pass
    
  @abstractmethod
  def all_tables(self, **kwargs):
    pass
    
  # other details omitted
`
(lives inside `generic_database.py`)

`
class MSSQLQueries(Queries):
  def ping(self):
    return "SELECT 1+1"
    
  def all_tables(self, **kwargs):
     return f"SELECT table_name FROM information_schema.tables WHERE TABLE_SCHEMA = '{schema}'"
     
   # other details omitted
`

(lives inside `mssql.py`)

Reply:
Done